src/PaginationCursor.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EventSauce\EventSourcing;
+
+use RuntimeException;
+use Throwable;
+
+use const JSON_THROW_ON_ERROR;
+
+final class PaginationCursor
```

Comment:
Why do we need a complex PaginationCursor with strings and integers? What are the envisioned use cases for it? Wouldn't it be more practical to have a PaginationCursor interface and a IntegerOffset or DateTimeOffset etc implementing it? 🤔

Reply:
I don't see overly compelling reasons to go either way. One thing about this more generic cursor is that it has serialization backed in, which is convenient. I can anticipate people wanting an abstract cursor if it would be an interface, which is _meh_. This map construct would allow people to do some type of filtering while retaining the same top-level API.

Reply:
Benefit for interface would be that it would be easier for repositories to expect a certain type. (since they implement an interface, they still have to do a check.) But at least they could easily check if the type of cursor is what it expects. 

Eg: 
```
class SomeRepositoryThatWantsASpecialCursor
{
    public function paginate(int $perPage, PaginationCursor|null|SpecialCursor $cursor = null): Generator
    {
        if($cursor !== null && !$cursor instanceof SpecialCursor){
            throw new Exception("SpecialCursor required");
        }
    }
}
```

With the final cursor, there is no way to check if the cursor is of a specific type. Meaning we'd have to check for odd cases in our repo's when requiring more complex cursors. 

The PaginationCursor interface would need a `toString(): string` and `fromString(string $string): static` method I guess.

Reply:
For me, this is not a significant benefit. In either case, the input and data will diverge and check against it will need to be made. I've modelled it currently like I did the message headers, where the data is relatively anonymous, but the vehicle of which is generic. I don't see any functional limitations due to either approach, and then I'm favouring the one I've proposed in the PR. If there are any concrete scenarios to discuss, please raise them, but otherwise I'm not convinced that a different way to represent this construct has significant upsides.

Reply:
Just to play around with it, I'll implement it in the way you've proposed and see what it's like.

Reply:
Took a few mins to implement it how I'd do it https://github.com/EventSaucePHP/EventSauce/pull/183

This PR has the concept of making cursor not nullable for pagination in repositories. Might reduce complexity in the repo's. Curious of your opinion on it :)